File: backend/app/models/localization.py
from datetime import datetime

from sqlalchemy import Column, DateTime, Integer, String, Text

from app.core.database import Base


class Localization(Base):
    # 本地化记录：保存市场、语言、策略和字幕处理结果。
    __tablename__ = "localizations"

    id = Column(Integer, primary_key=True, index=True)
    # project_id 用于把本地化结果绑定到短剧项目，当前先做弱关联，后续可升级为外键约束。
    project_id = Column(Integer, nullable=True, index=True)
    # episode_id / episode_no 用于把本地化版本绑定到具体短剧分集，当前采用弱关联以兼容旧数据。
    episode_id = Column(Integer, nullable=True, index=True)
    episode_no = Column(Integer, nullable=True, index=True)
    content_plan_id = Column(Integer, nullable=True)
    script_polish_id = Column(Integer, nullable=True)
    storyboard_id = Column(Integer, nullable=True)
    market = Column(String(100), nullable=False)
    language = Column(String(100), nullable=False)
    strategy = Column(String(100), nullable=False)
    result_json = Column(Text, nullable=False)
    created_at = Column(DateTime, default=datetime.now, nullable=False)

File: frontend/src/types/localization.ts
// 本地化处理请求：projectId + episodeId + sourceText 是本地化链路的核心上下文。
export type LocalizationProcessRequest = {
  projectId?: number | null
  episodeId?: number | null
  episodeNo?: number | null
  sourceText: string
  targetMarket: string
  targetLanguage: string
  strategy: string
  // 以下 snake_case 字段仅用于兼容旧历史数据和旧接口返回。
  market?: string
  language?: string
  source_text?: string | null
  project_id?: number | null
  episode_id?: number | null
  episode_no?: number | null
  contentPlanId?: number | null
  scriptPolishId?: number | null
  storyboardId?: number | null
}

// 单条字幕：覆盖翻译、配音、口型匹配和字幕处理状态。
export type LocalizedSubtitle = {
  index: number
  startTime: string
  endTime: string
  originalText: string
  directTranslation: string
  localizedText: string
  voiceStatus: string
  lipSyncStatus: string
  subtitleStatus: string
}

// 本地化流程节点：用于右侧流程卡片展示生产进度。
export type LocalizationWorkflowItem = {
  step: string
  status: string
  description: string
}

// 改写对比：用于说明本地化不是逐字翻译，而是在保留剧情含义的基础上做文化适配。
export type LocalizationComparisonItem = {
  originalText: string
  directTranslation: string
  localizedText: string
}

// 本地化处理结果：recordId 是当前持久化记录 ID。
export type LocalizationProcessResult = {
  recordId?: number
  projectId?: number | null
  episodeId?: number | null
  episodeNo?: number | null
  project_id?: number | null
  episode_id?: number | null
  episode_no?: number | null
  sourceText?: string
  targetMarket?: string
  targetLanguage?: string
  market: string
  language: string
  strategy: string
  subtitles: LocalizedSubtitle[]
  comparison: LocalizationComparisonItem[]
  workflowSteps: LocalizationWorkflowItem[]
  workflow?: LocalizationWorkflowItem[]
  status: string
}

// 本地化历史记录：保留输入条件、链路 ID 和处理结果；旧记录可能只有 market/language。
export type LocalizationHistoryItem = Omit<LocalizationProcessRequest, 'market' | 'language' | 'sourceText' | 'targetMarket' | 'targetLanguage'> & {
  id: number
  recordId?: number
  market: string
  language: string
  sourceText?: string
  targetMarket?: string
  targetLanguage?: string
  result: LocalizationProcessResult
  createdAt: string
}
